core/shared/src/main/scala/scalaz/zio/Semaphore.scala

```diff
@@ -4,112 +4,69 @@
-              if (requested > available) Left(((requested - available, gate), tail))
-              else Right(available - requested)
-            // n in queue but no more to release
-            case nonEmptyQueue if available == 0 => Left(nonEmptyQueue)
-          }
-
-        val updated: State = old match {
-          case Left(waiting) =>
-            releaseRecursively(waiting, toRelease)
-          case Right(available) =>
-            Right(available + toRelease)
-        }
-
-        ((old, updated), updated)
-      }.flatMap {
-        case (previous, now) =>
-          previous match {
-            case Left(neq) =>
-              val newSize = now match {
-                case Left(newNeq) => newNeq.size
-                case Right(_)     => 0
-              }
-              val released = neq.size - newSize
-              neq.take(released).foldRight(IO.unit) { (entry, unit) =>
-                openGate(entry) *> unit
-              }
-            case Right(_) => IO.unit
-          }
-      }
-
-  private def mkGate: IO[Nothing, Promise[Nothing, Unit]] = Promise.make[Nothing, Unit]
-
-  private def awaitGate(entry: (Long, Promise[Nothing, Unit])): IO[Nothing, Unit] =
-    IO.unit.bracket0[Nothing, Unit] { (_, useOutcome) =>
-      useOutcome match {
-        case _: Terminated[Nothing, Unit] => // Terminated outcome means either interruption or uncaught exception
-          state.update {
-            case Left((current, rest)) =>
-              // if entry is NonEmpty's head and queue is empty, swap to Right, but without any permits
-              if (current == entry && rest.isEmpty) Right(0)
-              // if entry is NonEmpty's head and queue is not empty, just drop this entry from head position
-              else if (current == entry && rest.nonEmpty) Left(rest.dequeue)
-              // this entry is not current NonEmpty's head, just drop it from queue
-              else Left((current, rest.filter(_ != entry)))
-
-            case Right(m) => Right(m)
-          }.void
-        case _ =>
-          IO.unit
-      }
-    }(_ => entry._2.get)
-
-  private def openGate[E](entry: (Long, Promise[E, Unit])): IO[E, Unit] =
-    entry._2.complete(()).void
-
-  private def count_(state: State): Long = state match {
-    case Left((head, tail)) => -(head._1 + tail.map(_._1).sum)
-    case Right(available)   => available
+  final def acquireN(requested: Long): IO[Nothing, Unit] = {
+    val acquire: (Promise[Nothing, Unit], State) => (IO[Nothing, Boolean], State) = {
+      case (p, Right(n)) if n >= requested => p.complete(()) -> Right(n - requested)
+      case (p, Right(n))                   => IO.now(false)  -> Left(IQueue(p -> (requested - n)))
+      case (p, Left(q))                    => IO.now(false)  -> Left(q.enqueue(p -> requested))
+    }
+
+    val release: (Boolean, Promise[Nothing, Unit]) => IO[Nothing, Unit] = {
+      case (_, p) =>
+        p.poll.void <> state.update {
```

Comment:
`p.poll.void` can have no effect, why use it?

Reply:
I was going to say the promise can only be completed if it is taken out of the queue, but interruption can occur between the `complete` call and the moment the queue is updated in the ref.

Reply:
I still don't understand. I think if you delete `p.poll.void <>`, the meaning of this code won't change.

Reply:
In the happy path, the promise given to `release` is already completed (that is part of `Promise.bracket`, specifically the `p.get`), which can only happen if it is not in the queue (that is imposed by the definition of `acquireN`/`releaseN`).

Actually I take back my previous comment, because `Promise.bracket` guarantees atomicity, _if_ the promise is completed we can deduce that it is not in the queue anymore! So, with `p.poll.void` we avoid the extra state update if the promise has already been removed from the queue.

Reply:
What does `p.poll` do? AFAIK, it doesn't do anything, it just peeks in the promise and returns the value as a result, which is then thrown away with `void`. Am I missing something?

Reply:
In this case, I'm only interested to know whether the promise is complete or not: and `poll` fails in the latter case, which means I can run provide an alternative action (update the state) with the `<>` combinator.

Reply:
Ah, OK, I see now that `poll` fails with `Unit` if the result is not available. Unfortunately there's a race condition—it's possible the promise will be completed between `p.poll.void` and `state.update`. Because other threads may be concurrently modifying state (releasing, which could trigger promise completion). If there's a race condition then we should always update the state. Or am I missing something else?

Reply:
The `release` action is inside a finaliser (via `io.ensuring(...)`), which means it is not interruptible and so this should be fine. Amirite?

Reply:
It's not interruptible, however, the state of the promise can still change during the execution of the action. `Promise.bracket` provides no guarantees on concurrent access, only a guarantee if acquire succeed, that release will succeed, regardless of interruption.

Reply:
That is correct 😞 Thanks for the explanation!